pkg/cmd/cli/secrets/unlink_secret_from_object.go

```diff
@@ -0,0 +1,114 @@
+Unlink (detach) secrets from a ServiceAccount
+If a secret is no longer valid for a pod, build or image pull, you may unlink it from a service account.`
+
+	unlinkSecretExample = ` # Unlink a secret currently associated with a service account like this:
+%[1]s serviceaccount/sa-name secret/secret-name secret/another-secret-name ...`
+)
+
+type UnlinkSecretOptions struct {
+	SecretOptions
+}
+
+// NewCmdUnlinkSecret creates a command object for detaching one or more secret references from a service account
+func NewCmdUnlinkSecret(name, fullName string, f *kcmdutil.Factory, out io.Writer) *cobra.Command {
+	o := &UnlinkSecretOptions{SecretOptions{Out: out}}
+
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s serviceaccount/sa-name secret/secret-name [secret/another-secret-name] ...", name),
+		Short:   "Detach secrets from a ServiceAccount",
+		Long:    unlinkSecretLong,
+		Example: fmt.Sprintf(unlinkSecretExample, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			if err := o.Complete(f, args, nil); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+			if err := o.Validate(); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := o.UnlinkSecrets(); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+
+		},
+	}
+
+	return cmd
+}
+
+func (o UnlinkSecretOptions) UnlinkSecrets() error {
+	serviceaccount, err := o.GetServiceAccount()
+	if err != nil {
+		return err
+	}
+
+	if err = o.unlinkSecretsFromServiceAccount(serviceaccount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// unlinkSecretsFromServiceAccount detatches pull and mount secrets from the service account.
+func (o UnlinkSecretOptions) unlinkSecretsFromServiceAccount(serviceaccount *kapi.ServiceAccount) error {
+	// All of the requested secrets must be present in either the Mount or Pull secrets
+	// If any of them are not present, we'll return an error and push no changes.
+	rmSecrets, err := o.GetSecrets()
+	if err != nil {
+		return err
+	}
+	rmSecretNames := o.GetSecretNames(rmSecrets)
+
+	var found bool
+
+	for _, secretname := range rmSecretNames.List() {
+		found = false
+
+		// Check the mount secrets
+		for i := len(serviceaccount.Secrets) - 1; i >= 0; i-- {
+			if secretname == serviceaccount.Secrets[i].Name {
+				found = true
+
+				// Unlink this entry from Secrets
+				serviceaccount.Secrets = append(serviceaccount.Secrets[:i], serviceaccount.Secrets[i+1:]...)
+			}
+		}
+
+		// Check the image pull secrets
+		for i := len(serviceaccount.ImagePullSecrets) - 1; i >= 0; i-- {
+			if secretname == serviceaccount.ImagePullSecrets[i].Name {
+				found = true
+
+				// Unlink this entry from ImagePullSecrets
+				serviceaccount.ImagePullSecrets = append(serviceaccount.ImagePullSecrets[:i], serviceaccount.ImagePullSecrets[i+1:]...)
+			}
+		}
+
+		if !found {
+			// This secret didn't appear in either secret type
+			return fmt.Errorf("%s is not associated with service account %s",
```

Comment:
If you are going to error on deletions where the thing doesn't exist, you need a `--force` flag so that people can script deletion.  Generally we don't error on deletions that don't exist today, so this would be inconsistent with our other commands.